include/boost/wave/util/cpp_macromap.hpp

```diff
@@ -508,6 +530,53 @@ macromap<ContextT>::is_defined(string_type const &str) const
     return is_defined(str, cit, 0);
 }
 
+template <typename ContextT>
+bool
+macromap<ContextT>::is_space(char c)
+{
+    // leave out CR/LF to be consistent with on_include()
+    return ((c == ' ') || (c == '\t') || (c == '\v') || (c == '\f'));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+//  has_include(): returns whether a path expression is an valid include file
+//
+///////////////////////////////////////////////////////////////////////////////
+template <typename ContextT>
+template <typename IteratorT>
+inline bool
+macromap<ContextT>::has_include(
+    IteratorT const &begin, IteratorT const &end,
+    bool is_quoted_filename, bool is_system) const
+{
+    typename ContextT::token_sequence_type filetoks;
+
+    if (is_quoted_filename) {
+        filetoks = typename ContextT::token_sequence_type(begin, end);
+    } else {
+        IteratorT first = begin;
+        IteratorT last = end;
+        ctx.expand_whole_tokensequence(first, last, filetoks);
+    }
+
+    // extract tokens into string
+    std::string fn(boost::wave::util::impl::as_string(filetoks).c_str());
+
+    // trim leading and trailing blanks
+    boost::algorithm::trim_if(fn, is_space);
```

Comment:
What about (c-style) comments? Is it a good idea to do this on the character level? It might be better to do the trimming on the token level.

Reply:
I'm following the pattern [here](https://github.com/boostorg/wave/blob/bacb94f394943f61517406ad0136093bebdc8158/include/boost/wave/util/cpp_iterator.hpp#L1722), where includes are handled by combining the tokens, then trimming. Could there be comments inside a `__has_include()`?

Reply:
> I'm following the pattern here, 

Should you use the existing `trim_whitespace()` in this case?

> where includes are handled by combining the tokens, then trimming.

Not sure if comments are removed by the parser in this case. I need to check.

> Could there be comments inside a __has_include()?

Sure, why not, it's equivalent to whitespace after all?

Reply:
Yeah, so... the weird thing is the existing trim code is in an internal namespace in `cpp_iterator.hpp`, and the macro expansion code is in `cpp_macromap.hpp`. I guess this is a side effect of how `__has_include` crosses the existing boundaries between functionality, combining a little "include processing" and a little "macro processing". I struggled with this for a while. You could say that this approach (a minor wheel reinvention) was the one that I disliked the least :) What would you suggest?

I will check how clang/gcc handle this situation.

Reply:
Couldn't we move the helper code to some commonly included spot (for instance, here: https://github.com/boostorg/wave/blob/develop/include/boost/wave/util/macro_helpers.hpp)?

Reply:
What if we move the whitespace trimming into util/macro_helpers.hpp? It has similar functionality.

Reply:
That sounds like a great idea :) (sorry missed your previous message somehow)

Reply:
I can confirm that gcc handles comments within `__has_include()` and Wave... does not, yet. Thanks for the observation.

Reply:
Well, that was an easy fix. Comments were removed by the skipper already - I just needed to fix a check I added.

Reply:
Any other concerns at the moment? I'll forge ahead with unit tests in the meantime.